client/app/src/test/java/com/example/criengine/RequestsForBookAdapterTest.java

```diff
@@ -0,0 +1,146 @@
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+
+import java.util.ArrayList;
+
+@RunWith(RobolectricTestRunner.class)
+public class RequestsForBookAdapterTest {
+    @Rule public MockitoRule mockitoRule = MockitoJUnit.rule();
+
+    @Mock
+    DatabaseWrapper dbwMock;
+
+    DatabaseWrapper dbw;
+    private Context context;
+
+    String requester;
+    Book book;
+
+    View adaptedView;
+    Button rejectedButton;
+    Button acceptButton;
+    TextView usernameTextView;
+
+    /**
+     * Initializes a mock database and context before every test
+     */
+    @Before
+    public void setup() {
+        dbwMock.userId = "testUser";
+        dbw = new DatabaseWrapper(dbwMock);
+
+        context = ApplicationProvider.getApplicationContext();
+
+        book = new Book();
+
+        requester = "";
+    }
+
+    /**
+     * Initializes a new RequestsForBookAdapter with the given list
+     *  and obtains the view of the first item
+     *  It then it properties: adaptedView, actionButton and statusText
+     */
+    public void setupAdaptedItemView(ArrayList<String> list) {
+        RequestsForBookAdapter RequestsForBookAdapter =
+                new RequestsForBookAdapter(context, list, book);
+        View itemView = View.inflate(context, R.layout.user_accept_or_reject, null);
+        adaptedView = RequestsForBookAdapter.getView(0, itemView, null);
+        usernameTextView = adaptedView.findViewById(R.id.user_name);
+        rejectedButton = adaptedView.findViewById(R.id.user_reject);
+        acceptButton = adaptedView.findViewById(R.id.user_accept);
+    }
+
+
+    /**
+     * Initializes a new RequestsForBookAdapter with a list containing only the requester property
+     *  and obtains the view of the first item
+     *  It then it properties: adaptedView, actionButton and statusText
+     */
+    public void setupAdaptedItemView() {
+        ArrayList<String> requesterList = new ArrayList<>();
+        requesterList.add(requester);
+        setupAdaptedItemView(requesterList);
+    }
+
+
+    /**
+     * Tests Adapter constructor
+     */
+    @Test
+    public void ConstructorTest() {
+        ArrayList<String> requesterList = new ArrayList<>();
+        RequestsForBookAdapter RequestsForBookAdapter =
+                new RequestsForBookAdapter(context, requesterList, book);
+        Assert.assertNotNull(RequestsForBookAdapter);
+    }
+
+    /**
+     * Tests that the Requester username is displayed
+     */
+    @Test
+    public void RequesterUsernameTest() {
+        requester = "user1";
+        setupAdaptedItemView();
+        Assert.assertEquals(requester, usernameTextView.getText());
+    }
+
+    /*
+    @Test

```

Comment:
These tests were supposed to test the onclick handlers for the buttons in the adapter. But the onlclick listeners attempt to start a new Activity which robolectric very much doesnt like, especially since these tests are carried out without an Activity to begin with.

Reply:
robolectric has "shadow" objects that you're supposed to use, it doesn't let the activity actually carry out, only that it would have.
http://robolectric.org/writing-a-test/

Reply:
I saw that earlier but I wasn't able to get it to work. I could get it work it it later, but I'll leave it commented for now